llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -1121,6 +1121,40 @@ static void computeKnownBitsFromOperator(const Operator *I,
       break;
     }
 
+    Value *V;
+    // Handle bitcast from floating point to integer.
+    if (match(const_cast<Operator *>(I), m_ElementWiseBitCast(m_Value(V))) &&
+        V->getType()->isFPOrFPVectorTy()) {
+      KnownFPClass Result = computeKnownFPClass(V, fcAllFlags, Depth + 1, Q);
+      if (Result.SignBit) {
+        if (*Result.SignBit)
+          Known.makeNegative();
+        else
+          Known.makeNonNegative();
+      }
+
+      Type *FPType = V->getType()->getScalarType();
+      if (FPType->isIEEELikeFPTy()) {
```

Comment:
I think this should be implicit from getFPMantissaWidth returning -1, but I see that is not what it returns for ppcf128. I don't think that answer makes any sense. Can we just change that to treat ppcf128 as not understandable?

Reply:
It will cause assertion failure if we treat ppcf128 as not understandable. Some existing code calls `getFPMantissaWidth` unconditionally.

https://github.com/llvm/llvm-project/blob/61f1f1300254e91334404f0d22e75e96f0d98a85/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L7511-L7514

Reply:
I misread the code here, I think you don't need to check isIEEELikeFPTy here. PPC_FP128 isn't covered by it, which makes more sense

Reply:
Should we handle `x86_fp80`? `getFPMantissaWidth` returns 64 for `x86_fp80`. But fp80 doesn't follow the IEEE floating-point format :( 
https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format

Reply:
For x86_fp80, you still know all the exponent bits are 1. The mantissa bits are more confusing, but you're only really using the mantissa bits here to get the number of exponent bits

Reply:
> For x86_fp80, you still know all the exponent bits are 1 for inf. The mantissa bits are more confusing, but you're only really using the mantissa bits here to get the number of exponent bits

I prefer to just handle IEEE floating-point types.

Reply:
If you query the constant values instead of reverse engineering the bit layout, you would get it for free

Reply:
> If you query the constant values instead of reverse engineering the bit layout, you would get it for free

Done. A mistake about inf encoding was also fixed.